alf/environments/suite_carla.py

```diff
@@ -0,0 +1,1163 @@
+            logging.info("Stopping Carla server: %s" % command)
+            _exec(command)
+        if self._process:
+            self._process.kill()
+
+
+@gin.configurable
+class CarlaEnvironment(TorchEnvironment):
+    """Carla simulation environment.
+
+    In order to use it, you need to either download a valid docker image or
+    a Carla package.
+    """
+
+    def __init__(self,
+                 batch_size,
+                 map_name,
+                 vehicle_filter='vehicle.*',
+                 safe=True,
+                 step_time=0.05):
+        """
+        Args:
+            safe (bool): avoid spawning vehicles prone to accidents.
+            step_time (float): how many seconds does each step of simulation represents.
+        """
+        super().__init__()
+
+        with _get_unused_port(2000, n=2) as (rpc_port, streaming_port):
+            self._server = CarlaServer(rpc_port, streaming_port)
+
+        self._batch_size = batch_size
+        self._world = None
+        try:
+            for i in range(10):
+                try:
+                    logging.info(
+                        "Waiting for server to start. Try %d" % (i + 1))
+                    self._client = carla.Client("localhost", rpc_port)
+                    self._world = self._client.load_world(map_name)
+                    break
+                except RuntimeError:
+                    continue
+        finally:
+            if self._world is None:
+                self._server.stop()
+                assert self._world is not None, "Fail to start server."
+
+        logging.info("Server started.")
+        self._alf_world = World(self._world)
+        self._safe = safe
+        self._vehicle_filter = vehicle_filter
+
+        settings = self._world.get_settings()
+        settings.synchronous_mode = True
+        settings.fixed_delta_seconds = step_time
+
+        self._world.apply_settings(settings)
+        self._map_name = map_name
+        self._spawn()
+
+    def _spawn(self):
+        blueprints = self._world.get_blueprint_library().filter(
+            self._vehicle_filter)
+        assert len(
+            blueprints) > 0, "Cannot find vehicle '%s'" % self._vehicle_filter
+        if self._safe:
+            blueprints = [
+                x for x in blueprints
+                if int(x.get_attribute('number_of_wheels')) == 4
+            ]
+            blueprints = [x for x in blueprints if not x.id.endswith('isetta')]
+            blueprints = [
+                x for x in blueprints if not x.id.endswith('carlacola')
+            ]
+            blueprints = [
+                x for x in blueprints if not x.id.endswith('cybertruck')
+            ]
+            blueprints = [x for x in blueprints if not x.id.endswith('t2')]
+        assert len(
+            blueprints
+        ) > 0, "Cannot find safe vehicle '%s'" % self._vehicle_filter
+
+        spawn_points = self._world.get_map().get_spawn_points()
```

Comment:
Each map only has a fixed number of spawn points?

Reply:
correct.